### ISSUE TITLE
Softmax stucks the network

### ISSUE THREAD

**Issue by March-08** [FluxML/Flux.jl#1546]
When I declare a Network in this way and I use the softmax function in the output model, the loss does not decrease.
The network works fine without the softmax function.

```
function custom_net(input::Int64,hidden::Int64,output::Int64)
    linear1 = linear(input, hidden) # we can access linear1.W etc
    linear2 = linear(hidden, output)
    ps = Flux.params(linear1.W , linear1.b , linear2.W, linear2.b)
    model(x) =  softmax(linear2(σ.(linear1(x))))
    return model,ps
  end
```
```

m,ps = CustomizeNet.custom_net(2,2,1)

data,xtest,ytest = DataCostumizer.dummy_data_micronet(1000,100)
loss(x, y) = Flux.mse(m(x),y)
opt = ADAM()
initial_masks = [ones(2,2),ones(1,2)]

evalcb = () -> println("loss : ",loss(xtest,ytest))
@Flux.epochs 10 Flux.train!(loss, ps, data, opt,cb = throttle(evalcb, 10)) 
```

```
[ Info: Epoch 1
loss : 0.28823234186242325
[ Info: Epoch 2
loss : 0.28823234186242325
[ Info: Epoch 3
loss : 0.28823234186242325
[ Info: Epoch 4
loss : 0.28823234186242325
[ Info: Epoch 5
loss : 0.28823234186242325
[ Info: Epoch 6
loss : 0.28823234186242325
[ Info: Epoch 7
loss : 0.28823234186242325
[ Info: Epoch 8
loss : 0.28823234186242325
[ Info: Epoch 9
loss : 0.28823234186242325
[ Info: Epoch 10
loss : 0.28823234186242325
```

**Comment by darsnack** [comment]
What is `data`, `xtest`, and `ytest`? You're using a softmax output with a MSE loss which is an unusual combination. MSE is usually used in regression problems where the output is an arbitrary real vector. Softmax is used in classification problems where the output is one-hot. Using softmax in your context would only work if the labels of your data are one-hot.

**Comment by darsnack** [comment]
Here's a quick comparison of the gradient through just softmax for arbitrary real vectors as labels vs. one-hot vectors:
```julia
julia> gradient(x -> Flux.mse(softmax(x), rand(100, 32)), rand(100, 32))
([-1.2562316294644186e-6 -7.892260049466905e-7 … -2.607503530280089e-6 -2.0693881073535094e-6; 1.5160567348431414e-6 1.8368543513887875e-6 … -1.9568047655421484e-6 -2.3445055993973654e-6; … ; 7.611966951807486e-7 -2.9344330016232494e-8 … -2.1984512772476406e-6 -3.010328378730786e-6; 2.845061403825191e-6 1.443615074232501e-6 … -1.922457791407221e-6 -3.241696793290808e-7],)

julia> gradient(x -> Flux.crossentropy(softmax(x), Flux.onehotbatch(rand(1:100, 32), 1:100)), rand(100, 32))
([0.00019456963152819304 0.00024611430129795677 … 0.00018988055347552517 0.0004550984662941538; 0.0004512709836350396 0.00024707636932591294 … 0.0004383246345834223 0.0002801790912957242; … ; 0.00021483703394882355 0.00033688965051615835 … 0.00046588090936694335 0.00030580823542496237; 0.0005172598306390038 0.00020861994532408673 … 0.0004099735783934692 0.00019565344611177293],)
```

You can see that when using MSE with softmax and arbitrary labels, the gradients are orders of magnitude smaller.

**Comment by March-08** [comment]
yes you're right, I was about to change the loss, but still the should change a bit at each epoch, shouldn't it?

 ```
function dummy_data_micronet(train::Int64, test::Int64)
        xtrain = rand(train,2) |> cpu
        ytrain = rand(train) |> cpu
        xtest = rand(test,2) |>cpu
        ytest = rand(test) |> cpu
        data = [(xtrain', ytrain')]  |> cpu

        return data,xtest',ytest'
    end
```

**Comment by darsnack** [comment]
Softmax is designed to create outputs that look like probability vectors. If you labels don't match that, then functionally speaking, your model is not capable of expressing the output space that your labels lie. In other words, your model is only able to output a subset of real vectors, and your labels do not lie completely within that subset. So, there's probably some lower bound on the loss that this model is able to achieve given MSE + your labels. My guess is that the problem is so simple (especially with the dimensionality of your input/output), that it achieves that lower bound within the first epoch, and it never changes after that.

**Comment by darsnack** [comment]
Either way, the gradients for softmax are clearly non-zero when the function is used appropriately, so I think it is safe to say that this is not a bug with softmax. I'm going to close the issue, but please continue to comment if you face problems.

**Comment by darsnack** [comment]
Actually, I need to amend my earlier comment. What I wrote is true and it will probably hold for higher dimensional output. But in your posted example, the issue is even worse since the output is 1D. So the softmax applied to your output will always be 1.
```julia
julia> softmax(rand(1))
1-element Array{Float64,1}:
 1.0
```

**Comment by March-08** [comment]
ok, I got it, thank you for your answer!
So I am assuming that the binarycross entropy should be more appropriate in my case, but I still got an error:

`ERROR: output an array, so the gradient is not defined. Perhaps you wanted jacobian.`

**Comment by March-08** [comment]
you're right! perfect, I have to pay more attention even in these tiny testing networks!
thank you so much.
